pkg/applicationserver/applicationserver.go

```diff
@@ -1352,6 +1414,28 @@ func (as *ApplicationServer) handleDownlinkNack(ctx context.Context, ids *ttnpb.
 			}
 
 			items := []*ttnpb.ApplicationDownlink{msg}
+
+			if msg.Confirmed {
+				if msg.ConfirmedRetry == nil {
+					msg.ConfirmedRetry = &ttnpb.ApplicationDownlink_ConfirmedRetry{
+						Attempt: 0,
+					}
+				}
+				msg.ConfirmedRetry.Attempt++
+
+				maxRetries := uint32(as.config.Downlinks.ConfirmationConfig.DefaultRetryAttempts)
+				if msg.ConfirmedRetry.MaxAttempts.GetValue() > 0 {
+					maxRetries = msg.ConfirmedRetry.MaxAttempts.GetValue()
```

Comment:
Can use `.Value` directly, and you may consider having a variable directly in the `if` definition, like `if maxAttempts := msg.ConfirmedRetry.MaxAttempts; maxAttempts != nil`.